EDP-380C-28-R2-Lab3.R

```diff
@@ -12,3 +12,411 @@ source("scripts/rmvnorm.R")
+  
+  # Method 1
+  
+  ## Drawing X 
+  
+  p <- length(p_x$mu)
+  x <- with(p_x, rmvnorm(n, mu, Sigma, rho))
+  
+  beta <- with(p_y, matrix(beta, nrow = p))
+  eq23 <- with(p_y, t(beta) %*% cov(x) %*% beta * (1/r2 -1))
+  sigma_e <- sqrt(eq23)
+  beta0 <- with(p_y, mu - (t(p_x$mu) %*% beta))
+  
+  y1 <- with(p_y, (matrix(1, n) %*% beta0)+ x %*% beta + rnorm(n, 0, sigma_e))
+  
+  output <- cbind(y1,x)
+  attr(output, 'par_xy') <- list(p_x, p_y)
+  return(output) 
+}
+
+## Parameters 
+set.seed(6972)
+n <- 100000
+p_x <- list(mu = rep(0, 5),
+            Sigma = c(1, sqrt(2), sqrt(3), 2, sqrt(5)), 
+            rho = 0.15)
+
+p_y <- list(beta = rep(1, 5), mu = 25, r2 = 0.5)
+
+m1 <- gen_m1(n, p_x, p_y)
+
+m_3d1 <- OLS(m1[,1], cbind(m1[,2:6]))
+
+# Estimate          SE    t value Pr(>|t|)
+# b0    25.0219149 0.015222210 1643.77672        0
+# b1     0.9812633 0.015651423   62.69483        0
+# b2     1.0199224 0.011081135   92.04132        0
+# b3     0.9881525 0.009059712  109.07107        0
+# b4     0.9852174 0.007842738  125.62161        0
+# b5     1.0050576 0.007055326  142.45375        0
+# SD(e)  4.8135563          NA         NA       NA
+# R2     0.4991145          NA         NA       NA
+
+#### Answer 
+ans_3d1 <- rbind(b1 = cbind(p_y$beta[1], m_3d1[2, 1], p_y$beta[1] - m_3d1[2, 1]),
+                 b2 = cbind(p_y$beta[2], m_3d1[3, 1],p_y$beta[2] - m_3d1[3, 1]),
+                 b3 = cbind(p_y$beta[3], m_3d1[4, 1],p_y$beta[3] - m_3d1[4, 1]),
+                 b4 = cbind(p_y$beta[4], m_3d1[5, 1],p_y$beta[4] - m_3d1[5, 1]),
+                 b5 = cbind(p_y$beta[5], m_3d1[6, 1],p_y$beta[5] - m_3d1[6, 1]),
+                 rsq = cbind(p_y$r2, m_3d1[8, 1], p_y$r2 - m_3d1[8, 1]),
+                 mu = cbind(p_y$mu, mean(m1[,1]),p_y$mu - mean(m1[,1])))
+
+colnames(ans_3d1) <- c("Population", "Estimate", "Difference")
+rownames(ans_3d1) <- c("b1", "b2", "b3", "b4", "b5", "R2", "mu")
+
+
+# Population   Estimate    Difference
+# b1        1.0  0.9812633  0.0187366927
+# b2        1.0  1.0199224 -0.0199223718
+# b3        1.0  0.9881525  0.0118474752
+# b4        1.0  0.9852174  0.0147825927
+# b5        1.0  1.0050576 -0.0050575912
+# R2        0.5  0.4991145  0.0008855123
+# mu       25.0 25.0334707 -0.0334707122
+
+## Generating Method 2 
+gen_m2 <- function(n, p_x, p_y){
+  
+  
+  # Method 2
+  p <- length(p_x$mu)
+  x <- with(p_x, rmvnorm(n, mu, Sigma, rho))
+  
+  x_cor_mat <- with(p_x, matrix(rho, length(Sigma), length(Sigma)))
+  diag(x_cor_mat) <- 1
+  cor_mat <- with(p_y, rbind(c(1, rho), cbind(rho, x_cor_mat)))
+  Sigma <- diag(c(p_y$sd, p_x$Sigma)) %*% cor_mat %*% diag(c(p_y$sd, p_x$Sigma))
+  
+  
+  ## Beta 
+  beta <- solve(Sigma[2:(p+1), 2:(p+1)]) %*% Sigma[2: (p +1), 1]
+  
+  ## Residual Variance
+  eq26 <- with(p_y, sd^2 - t(beta) %*% Sigma[2:(p + 1), 1])
+  se <- sqrt(eq26)
+  
+  ## Rsquared 
+  rsqn <- cor_mat[1, 2:(p +1)] %*% solve(cor_mat[2:(p + 1), 2: (p + 1)]) %*% cor_mat[2:(p+1), 1]
+  
```

Comment:
Similar to before, it is much better to do `[-1, 1]`, `[-1, -1]`, etc. for the indexing.